yarpl/include/yarpl/flowable/FlowableOperator.h

```diff
@@ -466,6 +466,146 @@ class FromPublisherOperator : public Flowable<T> {
 };
 
+template <typename T, typename R>
+class FlatMapOperator : public FlowableOperator<T, R, FlatMapOperator<T, R>> {
+  using ThisOperatorT = FlatMapOperator<T, R>;
+  using Super = FlowableOperator<T, R, ThisOperatorT>;
+
+ public:
+  FlatMapOperator(
+      Reference<Flowable<T>> upstream,
+      folly::Function<Reference<Flowable<R>>(T)> func)
+      : Super(std::move(upstream)), function_(std::move(func)) {}
+
+  void subscribe(Reference<Subscriber<R>> subscriber) override {
+    Super::upstream_->subscribe(make_ref<FMSubscription>(
+        this->ref_from_this(this), std::move(subscriber)));
+  }
+
+ private:
+  using SuperSubscription = typename Super::Subscription;
+  class FMSubscription : public SuperSubscription {
+   public:
+    FMSubscription(
+        Reference<ThisOperatorT> flowable,
+        Reference<Subscriber<R>> subscriber)
+        : SuperSubscription(std::move(flowable), std::move(subscriber)) {}
+
+    void onNextImpl(T value) override {
+      LOG(INFO) << "FMSubscription::onNextImpl(" << value << ")";
+      CHECK(!currentStreamSubscriber_ && pending_);
+
+      auto&& flatMapOp = this->getFlowableOperator();
+      Reference<Flowable<R>> newStream = flatMapOp->function_(std::move(value));
+      currentStreamSubscriber_ =
+          yarpl::make_ref<MappedStreamSubscriber>(this->ref_from_this(this));
+      newStream->subscribe(currentStreamSubscriber_);
+    }
+
+    void onCompleteImpl() override {
+      LOG(INFO) << "FMSubscription::onCompleteImpl()";
+      SuperSubscription::onCompleteImpl();
+    }
+
+    void onErrorImpl(folly::exception_wrapper ex) override {
+      LOG(INFO) << "FMSubscription::onErrorImpl()";
+      SuperSubscription::onErrorImpl(std::move(ex));
+    }
+
+    // called from MappedStreamSubscriber
+    void mappedOnNextImpl(R value) {
+      LOG(INFO) << "FMSubscription::mappedOnNextImpl(" << value << ")";
+      requested_--;
+      SuperSubscription::subscriberOnNext(std::move(value));
+    }
+
+    void mappedOnCompleteImpl() {
+      LOG(INFO) << "FMSubscription::mappedOnCompleteImpl()";
+
+      CHECK(!currentStreamSubscriber_);
+      CHECK(!pending_);
+
+      if (requested_) {
+        pending_ = true;
+        SuperSubscription::request(1);
+      }
+    }
+
+    void request(int64_t n) override {
+      LOG(INFO) << "FMSubscription::request(" << n << ")";
+      requested_ += n;
+
+      if (!currentStreamSubscriber_ && !pending_) {
+        LOG(INFO) << "FMSubscription::request(" << n
+                  << ") nothing pending, requesting!";
+        pending_ = true;
+        SuperSubscription::request(1);
+      } else {
+        CHECK(currentStreamSubscriber_);
+        LOG(INFO) << "FMSubscription::request(" << n
+                  << ") passing request through to currentStreamSubscriber_";
+        currentStreamSubscriber_->request(n);
+      }
+    }
+
+    void cancel() override {
+      LOG(INFO) << "FMSubscription::cancel()";
+      SuperSubscription::cancel();
```

Comment:
this should cancel all newStream subscriptions, not only the last one.

Reply:
n/a with updates